### DISCUSSION TITLE
What to do with `wire:model.lazy` in V3?

### DISCUSSION THREAD

**Discussion by calebporzio** [livewire/livewire#5743] [8. V3 (maintainers only to post)]
### V2 Behavior

In Livewire V2, there were 3 variations of `wire:model`:

```html
<input type="text" wire:model="search"> <!-- "live" --> 
<input type="text" wire:model.lazy="search"> <!-- "lazy" --> 
<input type="text" wire:model.defer="search"> <!-- "deferred" --> 
```

Each input above sends fewer network requests than the last:

* `wire:model` - send many requests as a user types
* `wire:model.lazy` - only send the request when the user tabs away from the input
* `wire:model.defer` - only send the request when the user hits "submit"

### V3 Behavior

In V3, `wire:model` is "deferred" by default:

```html
<input type="text" wire:model="search"> <!-- "deferred" --> 
<input type="text" wire:model.lazy="search"> <!-- "lazy" --> 
<input type="text" wire:model.live="search"> <!-- "live" --> 
```

Here's the new lineup:

* `wire:model` - only sends a request when the user presses "submit"
* `wire:model.lazy` - only send the request when the user tabs away from the input
* `wire:model.live` - sends requests as the user types

### The problem

Because `wire:model` in V2 updated the server as the user typed, `.lazy` felt like a good name for "tab away" behavior for two reasons:

1. It is "lazier" than the default
2. VueJS and Alpine use `v-model.lazy` and `x-model.lazy` for only updating the value on the `blur` event (which gets fired when tabbing away)

However, now that `wire:model` is NOT aggressive by default, adding `.lazy` actually makes `wire:model` MORE aggressive; the opposite of "lazier".

### Solutions

A) Stick with `.lazy` even though it's counter-intuitive because it's a convention

B) Change `.lazy` to `.blur`

C) Find a better word than `.blur`

I think it's worth changing (while maybe supporting lazy for backwards compatibility). However, I'm not sure `.blur` is intuitive enough for people.

**Comment by tljohansson** [comment]
According to my taste, I believe blur is appropriate. Would also be the same usage as AlpineJS, which makes more sense.

**Reply by adevade** [reply]
Yeah, I agree! And the same as the native JS `blur` event.

> The blur event fires when an element has lost focus.
> _Source: [MDN](https://developer.mozilla.org/en-US/docs/Web/API/Element/blur_event)_

**Comment by danharrin** [comment]
I think either `.blur` or something else other than `.lazy`. You're right that it is inconsistent right now. And it doesn't matter that it isn't the same as Vue or Alpine, as `x-model` is not the same behaviour as `wire:model` either since v3. So it's best to just make this tool consistent with itself rather than others.

I don't think `.blur` is perfect, but I can't think of a better one-word one, so that's where my vote is.

**Reply by Jubeki** [reply]
Some other words instead of blur:
- later
- unfocus
- leave
- away

**Reply by calebporzio** [reply]
Good points Dan, I agree

**Reply by danharrin** [reply]
Here is my raw feedback for those suggestions @Jubeki 
- `later` is not specific enough as to when the sync happens
- `unfocus` is no better than `blur` IMO, and sounds clunkier
- `leave` and `away` are not used as much as `blur` and `unfocus` in JavaScript APIs in my experience, but maybe I am missing some good examples

**Comment by comes** [comment]
`<input name="foo" wire:model="bar" x-on:blur="$model().sync" />` or similar.
Otherwise I vote for `wire:model.lazy`.

**Reply by calebporzio** [reply]
Ah interesting alternative. Already have a the precedence set so gonna stick with the modifier. Good to know, thanks!

**Comment by Artessan** [comment]
So if .live has a modifier for timing .live.2s maybe allow something like live.lazy or  live.away

**Reply by Artessan** [reply]
Not a maintainer sorry, saw after the comment.

**Reply by danharrin** [reply]
I don't think `.live.lazy` is that bad! As long as `.live.debounce` is also a thing

**Reply by calebporzio** [reply]
`.live.lazy` is pretty interesting actually. Hmmmmm

**Reply by calebporzio** [reply]
What about this? `wire:model.live.blur`

I can support both `wire:model.blur` and `wire:model.live.blur`

I'm not sure `wire:model.live.lazy` feels great (saying "live" and "lazy" for the same thing)

**Reply by alxndrmlr** [reply]
^ FWIW this feels good to me re: `wire:model.live.lazy` I read that as "this is live and going to send a request to the server on change, but it will do it lazily (matching my other preconceptions about what lazy is).

Also given the breaking change, is it worth just forcing everything that is "live like" to hang of-off `wire:model.live.xyz` kinda like a namespace?

`wire:model` would then just have nothing hanging off it. The mental model becomes (the bare minimum is to wire:model) and it will work (via next network request) but if you want it to be reactive on user input events tack on `.live` and other modifiers `debounce` etc.

**Reply by calebporzio** [reply]
Yeah, this makes sense and is what I'm thinking too.

Good point re: .live.lazy not being that weird - tyty

**Reply by calebporzio** [reply]
Ok, I'm thinking this is the way and won't be any breaking changes.

`wire:model.lazy` will work for backwards compatibility

`wire:model.live.blur` is what will be documented

`wire:model.blur` will work too, but is less intuitive so it won't be documented

**Reply by ju5t** [reply]
Does this mean you will get `wire:model.live.debounce.1000ms` too?

**Reply by nam-co** [reply]
This would be great, even for Entangle().live.debounce.1000ms

**Comment by nuernbergerA** [comment]
I agree that `lazy` is a misleading term.
As Dan already pointed out, `wire:model` will already behave differently.

`blur` makes sense, maybe `tab` or `away` are better

**Comment by ju5t** [comment]
I've never seen `lazy` as the lazier option. Neither do I think of it like that in v3. I see it as:

- `wire:model` is very lazy in v3.
- `wire:model.lazy` is a little lazy.
- `wire:model.live` is not lazy at all.

So yea, I voted for keeping `lazy`.

**Comment by Hasnayeen** [comment]
Either `leave` or `away` or `tab` or `jump`, any of this make sense to me as that indicates what action being taken by the user causing the request, but if we think in terms of how frequently request are being sent then,

- `wire:model` - rarely sends request
- `wire:model.often` - sends request often
- `wire:model.live` - sends request continuously

though I'm not fan of thinking in terms of frequency.

**Comment by ahosker** [comment]
I would not overthink it. 
It's still 'lazier' than `.live` but not than the default,. `.blur` sounds fine, too (add an alias for `.lazy` temporarily).

**Comment by arhinful** [comment]
`.blur` makes sense as it'll follow JS blur(lost focus) event.

**Comment by PhiloNL** [comment]
I think `.blur` is the most consistent with the expected behavior (especially for people new to Livewire coming from the JS world). The `onblur` event (https://developer.mozilla.org/en-US/docs/Web/API/Element/blur_event) has been there for ages (`<input type="text" onblur="myFunction()">`). I see there's also `focusout` (https://developer.mozilla.org/en-US/docs/Web/API/Element/focusout_event), which is a bit more descriptive and could be an alternative.

**Reply by calebporzio** [reply]
Good calls Philo. Think I'm gonna go with: support `.lazy` for BC, and add `.blur`

`.blur` will actually behave slightly differently. It will actually listen for the `blur` event, where lazy listens for `change` instead of `input`

**Reply by nuernbergerA** [reply]
i would prefer to not have an alias
BC can be solved with a simple search `wire:model.lazy` and replace `wire:model.blur` job

**Comment by michaelconnelly** [comment]
`blur` 'feels' right and a lot of people will know its intention almost immediately.

**Comment by jacksleight** [comment]
Just to throw another option into the mix: `.eager` could work (like lazy vs. eager loading).

**Comment by nfauchelle** [comment]
https://laravel-livewire.com/docs/2.x/properties#debouncing-input
Is debounce staying, or that getting removed?

**Comment by johannesobier** [comment]
What I found out: Currently wire:model.blur only works on blur. in several projects I've tables with column search fields in the second (after the titles row). Instead of that wire:model.lazy stilll behaves in v2 in that hitting enter also sends a request. I used the migration tool. So now I have to change a lot of .blur modifiers back into .lazy. I din't find any mention of the behaviour of .lazy (being different to .blur) in the docs. Would be a good thing, to mention it there.

**Reply by danharrin** [reply]
Very good point there, the enter behaviour is useful.

**Reply by joshhanley** [reply]
@johannesobier I had a look into this, and `.lazy` works on the change event, not blur event. So the way you would do that now is `wire:model.change` not `.blur`.

https://github.com/livewire/livewire/blob/dc8ad6e08fa207944f26b8821776c156e88b460d/js/directives/wire-model.js#L41-L42

In V3 it has been renamed to `.change`. But `.lazy` still works as an alias for change. See here for details on lazy https://livewire.laravel.com/docs/wire-model#all-available-modifiers

Hope this helps!

**Reply by joshhanley** [reply]
Just double checked V2 and it also worked on the change event. So yeah `wire:model.change` is what you want in V3.

**Reply by johannesobier** [reply]
Thanks, wasn't aware of that

**Reply by joshhanley** [reply]
@johannesobier no worries! 🙂

**Reply by nam-co** [reply]
> Very good point there, the enter behaviour is useful.

I mention to Caleb the enter idea on X, a while ago but no luck, hope he reconsiders https://x.com/mlmnamco/status/1699823678173610487?s=46&t=QETBIoaAmhTxn7IKhN6FBQ
